static/js/services/update-user.js

```diff
@@ -1,145 +1,123 @@
-var _ = require('underscore');
-
 (function () {
 
   'use strict';
 
   var inboxServices = angular.module('inboxServices');
 
-  var getUserUrl = function(id) {
-    return '/_users/' + encodeURIComponent(id);
-  };
-
-  var removeCacheEntry = function($cacheFactory, id) {
-    var cache = $cacheFactory.get('$http');
-    cache.remove(getUserUrl(id));
-    cache.remove('/_users/_all_docs?include_docs=true');
-  };
-
   inboxServices.factory('UpdateUser',
     function(
-      $cacheFactory,
       $http,
-      $q,
-      $log,
-      DB
+      $location,
+      $log
     ) {
-
       'ngInject';
 
-      var createId = function(name) {
-        return 'org.couchdb.user:' + name;
-      };
 
-      var getOrCreateUser = function(id, name) {
-        if (id) {
-          return $http.get(getUserUrl(id), { cache: true, targetScope: 'root' })
-            .then(function(result) { return result.data; });
+      /**
+       * Uses the user api to store user updates.
+       *
+       * Updates are in the style of the /api/v1/users/{username} service, see
+       * its documentation for more details.
+       *
+       * If you pass a username as password this will be used as Basic Auth to
+       * the api (required for password updates). Otherwise it will use your
+       * existing cookie.
+       *
+       * @param      {String}  username       The user you wish to update, without org.couchdb.user:
+       * @param      {Object}  updates        Updates you wish to make
+       * @param      {String}  basicAuthUser  Optional username for Basic Auth
+       * @param      {String}  basicAuthPass  Optional password for Basic Auth
+       */
+      return function(username, updates, basicAuthUser, basicAuthPass) {
+        var url = '/api/v1/users/' + username;
+
+        if (basicAuthUser) {
+          url = [
+            $location.protocol(), '://', basicAuthUser, ':', basicAuthPass, '@',
+            $location.host(), ':', $location.port(), url
+          ].join('');
```

Comment:
I'm really uncomfortable with this. For one thing it appears to be injectable. Instead can we use `/api/v1/users/<username>` as the URL and provide the basic auth as a hashed header? An example of how to do this is on the [$http docs](https://docs.angularjs.org/api/ng/service/$http) under the "Setting HTTP Headers" title.

Reply:
Definitely injectable; e.g. `basicAuthUser` starting with a `/` and ending with a `?` will do it. Good catch.

Reply:
OK, so as discussed on Slack there isn't actually a security issue here (Unless @browndav can give a good example once he's awake).

I agree the code is gross, so I went to change it to use an Auth header, but interestingly enough that causes a new problem: if you get the password wrong, you get a browser-level authentication. I presume this happens with the header and not with the URL method because perhaps it considers it a different origin server?

Anyway, here are our options?:

Reply:
We're going to go with option 3

Reply:
Actually, we're not! Turns out I just needed to change the style of server error we're sending, and make sure to send `Accept: application/json` as a header and we're golden.